server/oauth.js

```diff
@@ -0,0 +1,299 @@
+						var parsedBody   = JSON.parse(body)
+						  , accessToken  = parsedBody.access_token
+						  , refreshToken = parsedBody.refresh_token
+						  , decoded      = jwtDecode(accessToken);
+
+						debug('Refreshed JWT received for:', decoded.email);
+
+						res.cookies.set('oAuthToken', accessToken);
+						res.cookies.set('oAuthRefreshToken', refreshToken);
+					} else {
+						debug('Failed to refresh token:', response.body);
+						// debug('Failed to refresh token:', response.statusCode);
+					}
+				});
+			}
+		}
+
+		// 2. Passthru to other middlewares.
+		next();
+	},
+	callbackMiddleware: function (req, res, next) {
+		var query = req.query;
+
+		if (req._parsedUrl.pathname !== '/eid-callback') {
+			next();
+			return;
+		}
+
+		debug('Callback received. Query:', query);
+
+		if (query.error) {
+			// TODO: How should we notify the user about the error? Flash message?
+			res.writeHead(302, { Location: '/' });
+			res.end();
+
+			return;
+		}
+
+		var state = req.cookies.get('OAuthState');
+
+		// Clear state as it's supposed to be one time only.
+		res.cookies.set('OAuthState', null);
+
+		if (state !== query.state) {
+			debug('State value did not match:', state, query.state);
+
+			// TODO: Notify user.
+			res.writeHead(302, { Location: '/' });
+			res.end();
+
+			return;
+		}
+
+		debug('States match');
+
+		request({
+			uri: env.oauth.tokenEndpoint,
+			method: 'POST',
+			headers: {
+				'Content-Type': 'application/x-www-form-urlencoded',
+				Authorization: 'Basic ' + new Buffer(env.oauth.clientId + ':'
+					+ env.oauth.clientSecret).toString('base64')
+			},
+			form: {
+				grant_type: 'authorization_code',
+				code: query.code,
+				redirect_uri: env.oauth.redirectUrl
+			}
+		}, function (error, response, body) {
+			if (error) {
+				debug('Error received from token endpoint:', error);
+
+				// TODO: Notify user.
+				res.writeHead(302, { Location: '/' });
+				res.end();
+			} else if (response.statusCode >= 200 && response.statusCode < 300) {
+				var parsedBody   = JSON.parse(body)
+				  , accessToken  = parsedBody.access_token
+				  , refreshToken = parsedBody.refresh_token
+				  , decoded      = jwtDecode(accessToken);
+
+				debug('JWT received for:', decoded.email);
+
+				userEmailMap(function (map) {
+					return map.get(serializeValue(decoded.email));
+				})(function (userId) {
+					if (userId) return userId;
+
+					var isPublicApp = req.$appName === 'public'
+					  , demoUserId  = isPublicApp ? null : res.cookies.get('demoUser');
+
+					if (!demoUserId) {
+						return createUser({
+							firstName: decoded.fname,
+							lastName: decoded.lname,
+							email: decoded.email,
+							roles: ['user']
+						});
+					}
+
+					return updateUser(demoUserId, {
+						isDemo: undefined,
+						firstName: decoded.fname,
+						lastName: decoded.lname,
+						email: decoded.email
+					});
+				}).done(function (userId) {
```

Comment:
Hmmm... so here we use promise inside a callback... from what I remember this is not the best practice. Maybe you could wrap request in a promise and chain the next promise with it?

Reply:
As request module does not conform to standard node way of doing callbacks (in respect to input arguments) this is proper (and working) way to do it. Nothing especially cringy about it (maybe to purists ;) ).